### ISSUE TITLE
YOLO11 Multi-Detection Head Networks

### ISSUE THREAD

**Issue by taytay-swift** [ultralytics/ultralytics#20265]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

I attempted to add an extra detection head to recognize new objects. During training, I froze the backbone as well as the original detection heads, with the aim of retaining the original object recognition capabilities while using the new detection head to identify new objects. However, during inference, the original detection heads failed to function properly.
How can I improve my network to achieve the desired outcome?

```
# Ultralytics 🚀 AGPL-3.0 License - https://ultralytics.com/license

# Ultralytics YOLO11 object detection model with P3/8 - P5/32 outputs
# Model docs: https://docs.ultralytics.com/models/yolo11
# Task docs: https://docs.ultralytics.com/tasks/detect

# Parameters
nc: 80 # number of classes
scales: # model compound scaling constants, i.e. 'model=yolo11n.yaml' will call yolo11.yaml with scale 'n'
  # [depth, width, max_channels]
  n: [0.50, 0.25, 1024] # summary: 181 layers, 2624080 parameters, 2624064 gradients, 6.6 GFLOPs
  s: [0.50, 0.50, 1024] # summary: 181 layers, 9458752 parameters, 9458736 gradients, 21.7 GFLOPs
  m: [0.50, 1.00, 512] # summary: 231 layers, 20114688 parameters, 20114672 gradients, 68.5 GFLOPs
  l: [1.00, 1.00, 512] # summary: 357 layers, 25372160 parameters, 25372144 gradients, 87.6 GFLOPs
  x: [1.00, 1.50, 512] # summary: 357 layers, 56966176 parameters, 56966160 gradients, 196.0 GFLOPs

# YOLO11n backbone
backbone:
  # [from, repeats, module, args]
  - [-1, 1, Conv, [64, 3, 2]] # 0-P1 1/2
  - [-1, 1, Conv, [128, 3, 2]] # 1-P2 1/4
  - [-1, 2, C3k2, [256, False, 0.25]]

  - [-1, 1, Conv, [256, 3, 2]] # 3-P3 1/8
  - [-1, 2, C3k2, [512, False, 0.25]]

  - [-1, 1, Conv, [512, 3, 2]] # 5-P4 1/16
  - [-1, 2, C3k2, [512, True]]

  - [-1, 1, Conv, [1024, 3, 2]] # 7-P5 1/32
  - [-1, 2, C3k2, [1024, True]]

  - [-1, 1, SPPF, [1024, 5]] # 9
  - [-1, 2, C2PSA, [1024]] # 10

# YOLO11n head
head:
  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 6], 1, Concat, [1]] # cat backbone P4
  - [-1, 2, C3k2, [512, False]] # 13

  - [-1, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 4], 1, Concat, [1]] # cat backbone P3
  - [-1, 2, C3k2, [256, False]] # 16 (P3/8-small)

  - [-1, 1, Conv, [256, 3, 2]]
  - [[-1, 13], 1, Concat, [1]] # cat head P4
  - [-1, 2, C3k2, [512, False]] # 19 (P4/16-medium)

  - [-1, 1, Conv, [512, 3, 2]]
  - [[-1, 10], 1, Concat, [1]] # cat head P5
  - [-1, 2, C3k2, [1024, True]] # 22 (P5/32-large)

  - [[16, 19, 22], 1, Detect, [80]] # 23 Detect(P3, P4, P5)

  - [22, 1, nn.Upsample, [None, 2, "nearest"]]
  - [[-1, 13], 1, Concat, [1]] # cat head P4
  - [-1, 2, C3k2, [512, False]] # 26 (P4/16-medium)

  - [[26], 1, Detect, [nc]] # 27 Detect(P3, P4, P5)

```

**Comment by UltralyticsAssistant** [comment]
👋 Hello @taytay-swift, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://reddit.com/r/Ultralytics) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

---

This is an automated response to help you get started quickly. An Ultralytics engineer will assist you soon 🛠️

**Comment by Y-T-G** [comment]
You can check this

https://y-t-g.github.io/tutorials/yolov8n-add-classes/

**Comment by taytay-swift** [comment]
> You can check this
> 
> https://y-t-g.github.io/tutorials/yolov8n-add-classes/

Thank you so much! Your solutions and blog posts are incredibly helpful for beginners like me! I've learned a lot from them. You're a fantastic engineer

**Comment by glenn-jocher** [comment]
I'm glad Y-T-G's resource helped you! For your multi-detection head approach with YOLO11, there are a few important considerations. The architecture you've created looks promising, but your freezing strategy might be causing issues. Instead of completely freezing the original detection head, try using a weighted loss approach where the original head has a smaller loss weight during training. This allows some adaptation while preserving most of its original capabilities.

Also, ensure you're properly handling both detection heads during inference by modifying the forward method in your model. The detection outputs need to be properly merged or handled separately depending on your use case.

If you need deeper customization with YOLO11's detection heads, reviewing the `head.py` module in the reference code would be beneficial as it contains the implementation details of how detection heads work.

**Comment by taytay-swift** [comment]
> I'm glad Y-T-G's resource helped you! For your multi-detection head approach with YOLO11, there are a few important considerations. The architecture you've created looks promising, but your freezing strategy might be causing issues. Instead of completely freezing the original detection head, try using a weighted loss approach where the original head has a smaller loss weight during training. This allows some adaptation while preserving most of its original capabilities.
> 
> Also, ensure you're properly handling both detection heads during inference by modifying the forward method in your model. The detection outputs need to be properly merged or handled separately depending on your use case.
> 
> If you need deeper customization with YOLO11's detection heads, reviewing the `head.py` module in the reference code would be beneficial as it contains the implementation details of how detection heads work.

Hi glenn-jocher, is your work still going well :)? I'm very happy to hear about a new method: the weighted loss method, but I'm not too sure about the details of the real test, do you have more study material for me? I will be very grateful.
Also I would like to ask you some questions I have: my usage scenario is as follows: I need to add a new class recognition to coco80 class recognition capability, if I use the weighted loss method, which form of my data dataset annotation instead?

A:
nc : 1
CLASSES:
- [my new class]

OR

B
nc:81
classes:
- person
... (original coco class)
- [my new class]

Translated with DeepL.com (free version)

**Comment by glenn-jocher** [comment]
Hello! Yes, everything's going well with our work - thanks for asking!

For implementing the weighted loss approach, you'd want option B - using a dataset with 81 classes that includes all original COCO classes plus your new class. This structure allows you to train on all classes simultaneously while applying different loss weights.

In your training configuration, you'd modify the loss weights to give less importance to the original detection head compared to your new one. You can accomplish this by customizing the `head.py` implementation to apply a smaller loss weight factor (e.g., 0.1-0.3) to the original head's loss while keeping full weight for your new head.

For merging detections during inference, you'll need to modify the forward method of your model to handle both detection heads and combine their outputs - either by concatenating results or using a post-processing function to merge predictions from both heads.

The YOLO11 detection head implementation details in the reference code (particularly in `head.py`) show how you can extend the `forward` method to handle multiple detection paths. If you're still facing challenges, the `Detect` and `YOLOEDetect` classes would be good starting points to understand how to customize this behavior.

**Comment by taytay-swift** [comment]
> Hello! Yes, everything's going well with our work - thanks for asking!
> 
> For implementing the weighted loss approach, you'd want option B - using a dataset with 81 classes that includes all original COCO classes plus your new class. This structure allows you to train on all classes simultaneously while applying different loss weights.
> 
> In your training configuration, you'd modify the loss weights to give less importance to the original detection head compared to your new one. You can accomplish this by customizing the `head.py` implementation to apply a smaller loss weight factor (e.g., 0.1-0.3) to the original head's loss while keeping full weight for your new head.
> 
> For merging detections during inference, you'll need to modify the forward method of your model to handle both detection heads and combine their outputs - either by concatenating results or using a post-processing function to merge predictions from both heads.
> 
> The YOLO11 detection head implementation details in the reference code (particularly in `head.py`) show how you can extend the `forward` method to handle multiple detection paths. If you're still facing challenges, the `Detect` and `YOLOEDetect` classes would be good starting points to understand how to customize this behavior.

I'm so glad to still be hearing from you, your messages have given me so much to learn from! Thank you very much for your time and effort :)

I would also like to ask you if you are aware of the distillation method, can he pass the learning ability of the original model to another model? As an official yolo developer, what do you think about this method, and is there any official yolo documentation for learning it?

**Comment by glenn-jocher** [comment]
Yes, knowledge distillation is indeed another excellent approach for your use case! Distillation allows you to transfer knowledge from a teacher model (your pretrained COCO model) to a student model (your new model with 81 classes).

The advantage of distillation for your scenario is that it can help the new model learn both the original COCO classes and your new class efficiently, while preserving the original model's detection capabilities. It works by training your student model to mimic the teacher's output distributions, not just the hard class predictions.

While Ultralytics doesn't have dedicated documentation specifically for distillation with YOLO models, we've seen community implementations that have shown promising results. The process typically involves:

1. Using your original COCO model as the teacher
2. Creating a student model with your 81-class structure
3. Computing a distillation loss that encourages the student to produce similar feature representations and predictions to the teacher for the original 80 classes

This approach can be more elegant than managing multiple detection heads and might give you better overall performance. If you're interested in exploring this route, examining the `torch.nn.functional.kl_div()` function would be helpful as it's commonly used for the distillation loss function in PyTorch implementations.
